File: apps/menu-bar/Sources/YTMMenuBarConnector/ConnectorProtocol.swift
import Foundation

enum ConnectorProtocol {
  static let hostName = "com.gormanity.ytm_enhancer.menu_bar"
  static let connectorId = "com.gormanity.ytm-enhancer.menu-bar"
  static let connectorName = "YTM Menu Bar"
  static let connectorVersion = AppMetadata.version
  static let protocolVersion = "1.0.0"
  static let uninstallRequestedType = "connector.uninstallRequested"
  static let permissions = [
    "playback:read",
    "playback:control",
    "track:read",
    "ytm:focus",
  ]

  static func hello(requestId: String) -> [String: Any] {
    [
      "type": "connector.hello",
      "requestId": requestId,
      "manifest": [
        "id": connectorId,
        "name": connectorName,
        "version": connectorVersion,
        "protocolVersion": protocolVersion,
        "permissions": permissions,
      ],
    ]
  }

  static func subscribePlayback(requestId: String) -> [String: Any] {
    [
      "type": "connector.subscribe",
      "requestId": requestId,
      "events": ["playback.state"],
    ]
  }

  static func playbackStateRequest(requestId: String) -> [String: Any] {
    [
      "type": "playback.getState",
      "requestId": requestId,
    ]
  }

  static func playbackAction(_ action: String, requestId: String) -> [String: Any] {
    [
      "type": "playback.action",
      "requestId": requestId,
      "action": action,
    ]
  }

  static func playbackSeek(time: Double, requestId: String) -> [String: Any] {
    [
      "type": "playback.seek",
      "requestId": requestId,
      "time": time,
    ]
  }

  static func focusYouTubeMusic(requestId: String) -> [String: Any] {
    [
      "type": "ytm.focus",
      "requestId": requestId,
    ]
  }
}

struct TrackMetadata: Decodable {
  let title: String?
  let artist: String?
  let album: String?
  let year: Int?
  let artworkUrl: String?
}

struct PlaybackState: Decodable {
  let title: String?
  let artist: String?
  let album: String?
  let year: Int?
  let artworkUrl: String?
  let nextTrack: TrackMetadata?
  let isPlaying: Bool
  let progress: Double
  let duration: Double
  let isShuffling: Bool?
  let repeatMode: String?
}

struct HostMessage {
  let type: String
  let requestId: String?
  let state: PlaybackState?
  let code: String?
  let message: String?

  init?(json: [String: Any]) {
    guard let type = json["type"] as? String else { return nil }
    self.type = type
    self.requestId = json["requestId"] as? String
    self.code = json["code"] as? String
    self.message = json["message"] as? String

    if let stateObject = json["state"] {
      self.state = try? JSONDecoder().decode(
        PlaybackState.self,
        from: JSONSerialization.data(withJSONObject: stateObject)
      )
    } else {
      self.state = nil
    }
  }
}

File: apps/windows-tray/src/YTMTray.Core/ConnectorProtocol.cs
using System.Reflection;

namespace YTMTray.Core;

public static partial class ConnectorProtocol
{
    public const string HostName = "com.gormanity.ytm_enhancer.tray";
    public const string ConnectorId = "com.gormanity.ytm-enhancer.tray";
    public const string ConnectorName = "YTM Tray";
    public static string ConnectorVersion =>
        typeof(ConnectorProtocol)
            .Assembly
            .GetCustomAttribute<AssemblyInformationalVersionAttribute>()
            ?.InformationalVersion ?? "0.1.4";
    public const string ProtocolVersion = "1.0.0";
    public const string UninstallRequestedType = "connector.uninstallRequested";

    public static readonly string[] Permissions =
    [
        "playback:read",
        "playback:control",
        "track:read",
        "ytm:focus"
    ];

    public static ConnectorHelloMessage Hello(string requestId) =>
        new(
            "connector.hello",
            requestId,
            new ConnectorManifest(
                ConnectorId,
                ConnectorName,
                ConnectorVersion,
                ProtocolVersion,
                Permissions
            )
        );

    public static ConnectorSubscribeMessage SubscribePlayback(string requestId) =>
        new("connector.subscribe", requestId, ["playback.state"]);

    public static PlaybackGetStateMessage PlaybackStateRequest(string requestId) =>
        new("playback.getState", requestId);

    public static PlaybackActionMessage PlaybackAction(string action, string requestId) =>
        new("playback.action", requestId, action);

    public static PlaybackSeekMessage PlaybackSeek(double time, string requestId) =>
        new("playback.seek", requestId, time);

    public static YtmFocusMessage FocusYouTubeMusic(string requestId) =>
        new("ytm.focus", requestId);
}

public sealed record ConnectorManifest(
    string Id,
    string Name,
    string Version,
    string ProtocolVersion,
    IReadOnlyList<string> Permissions
);

public sealed record ConnectorHelloMessage(
    string Type,
    string RequestId,
    ConnectorManifest Manifest
);

public sealed record ConnectorSubscribeMessage(
    string Type,
    string RequestId,
    IReadOnlyList<string> Events
);

public sealed record PlaybackGetStateMessage(string Type, string RequestId);

public sealed record PlaybackActionMessage(
    string Type,
    string RequestId,
    string Action
);

public sealed record PlaybackSeekMessage(
    string Type,
    string RequestId,
    double Time
);

public sealed record YtmFocusMessage(string Type, string RequestId);

public sealed record ConnectorSource(
    string Id,
    string Name,
    bool IsDevBuild,
    string? ExtensionId
)
{
    public string DisplayName => IsDevBuild ? $"{Name} (dev)" : Name;
}

public sealed record ConnectorErrorMessage(
    string Type,
    string? RequestId,
    string Code,
    string Message
);

public sealed record ActiveBrowserConnection(
    int ProcessId,
    DateTimeOffset ConnectedAt,
    ConnectorSource? Source
);

public sealed record TrackMetadata(
    string? Title,
    string? Artist,
    string? Album,
    int? Year,
    string? ArtworkUrl
);

public sealed record PlaybackState(
    string? Title,
    string? Artist,
    string? Album,
    int? Year,
    string? ArtworkUrl,
    TrackMetadata? NextTrack,
    bool IsPlaying,
    double Progress,
    double Duration,
    bool? IsShuffling,
    string? RepeatMode
);

public sealed class HostMessage
{
    public string Type { get; set; } = "";
    public string? RequestId { get; set; }
    public PlaybackState? State { get; set; }
    public ConnectorSource? Source { get; set; }
    public string? Code { get; set; }
    public string? Message { get; set; }
}

public static partial class ConnectorProtocol
{
    public static ConnectorErrorMessage AppBusy(ActiveBrowserConnection? connection) =>
        new("connector.error", null, "app_busy", AppBusyMessage(connection?.Source));

    private static string AppBusyMessage(ConnectorSource? source)
    {
        var browser = source?.DisplayName ?? "another browser";
        return $"YTM Tray is already connected to {browser}. Disconnect that browser before connecting here.";
    }
}
